### ISSUE TITLE
Project Diablo 2 Fork Support 

### ISSUE THREAD

**Issue by kvothed2** [fabd/diablo2-runewizard#36]
Hi @fabd !

First of all, thank you for your work, it has been incredibly useful and i'm very grateful for your endeavor! 

Here with Season 6 of PD2 coming up this Friday (Nov 25th), i'm trying to have this thing up and running (yet again :D). Due to my limited knowledge (i can read html / css, but have no clue about JS), i cannot understand why my github pages are failing to load.

In my fork i've updated Main branch with all the new and changed runeword-descriptions, runewords, item types. Changed a few things elsewhere to reflect PD2. In the gh-pages branch i synched it with yours, tried to do few updates to see if pages will work, but they don't. https://kvothed2.github.io/pd2-runewizard/
It is failing to load resource /src/main.ts
I tried to launch it from the gh-pages /root and /docs (no idea about this tho) and from main /root

If this is something you can help me with easily, i would be forever grateful. 
Thank you very much.
KvotheD

**Comment by fabd** [comment]
Quick question before I read your post throughly : did you manage to run it locally?

**Comment by fabd** [comment]
I'm a bad boy and I did not include a folder that I use for deployment.

I'm assuming you can run the site locally when you just type `vite`in terminal and after doing the [Local Development](https://github.com/fabd/diablo2-runewizard#local-development) basic steps (adding npm packages).

I see I did commit a `package-lock.json` so in theory at least you're running and building from the same packages - so that shouldn't be the issue.

**As for deployment** - so I am out of the loop and I forgot almost all of Jekyll but looking at [the Github Pages docs](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#about-publishing-sources) I think what I did was I used the default `docs/` folder as the "publishing source" - and then the way it works, is you push your site to a `gh-pages` branch, and Github will take everything that is INSIDE the `docs/` folder to publish.

When you do `vite build` your site is in the `dist/` folder.

I have a simple bash script to automate _part of_ the process - though not everything.

Copy/paste the following to a script in your local repo - name it `deploy.sh` make it executable `chmod +x deploy.sh` : 

```bash
#!/bin/bash

git co gh-pages

[[ $? -ne 0 ]] && {
  echo -e "\n  \033[30;43m TEST \033[0m ERROR! \033[0m \n"
  exit
}

echo -e "\n \033[32m ... reset gh-pages to main \033[0m \n"
git reset --hard main

echo -e "\n \033[32m... copy dist/ to docs/ \033[0m \n"
rm -rf docs/
mkdir -p docs
cp -rv dist/* docs/

echo -e "\n \033[32m... add .nojekyll file \033[0m \n"
touch .nojekyll

git add .nojekyll docs/*

echo -e "\n \033[32m... commit \033[0m \n"
git commit -m '(Pages): new build by `.fabd/deploy.sh`'

echo -e "\n  \033[30;42m OK! \033[0m  git push -f \n"
```

You might have to create the local branch with `git co -b gh-pages` 

Make sure your master is up to date, then `vite build`

Then run the script `./deploy.sh`

Now as the last line of the script say, you can either just type `git push -f ` : you're forcing the changes because we just overwrite agian and again the gh-pages branch, we don't care about losing changes - since it is a "clone" of the master, purely for deployment. 

Your URL https://kvothed2.github.io/pd2-runewizard/  will then reflect the contents of `docs/` within that `gh-pages` branch.

Soon after you did the push, go to your repo, and you'll see on the sidebar "Environments  > Github Pages" and sometims it says "deploying..." or something like that. If you click that link you can diuble-check on the [Deployments](https://github.com/kvothed2/pd2-runewizard/deployments/activity_log?environment=github-pages) page that your last deployment was complete.. usually takes a couple minutes.

Let me know if you got it working.

**Comment by fabd** [comment]
Just to be sure - you're going to need to **build** the site before deployment. Hopefully you got that part working. In theory at least if you follow the basic installation steps, you're on some linux distro with nodejs & "node packagae manager" installed - you should be able to build the site with `vite build`

If not we can try to work it out but you have to be very specific at what step you're stuck.

**Comment by fabd** [comment]
So as you can see the original repo does have neither a `docs` or a `dist` folder - and the `.gitignore` file specifically ignores the dist/ folder.

Tjhe `dist/` (distribution) is basically the compiled website, compiled CSS/HTML/JS .. that would be your root folder if you were to manually move the files to a server's "public_html/" folder with like a FTP prgoram.

Now on Github Pages , there is just this convention where they will basically assume your `docs/` folder in your branch, to be the root folder of your published site.

Hence, the deploy script above :   

* switches to `gh-pages`
* resets the branch - because we overwrite the last published version
* re-creates a `docs/` folder, and copies all of `dist/` to it
* then you push that branch
* Github Pages notices the update, and takes everything inside of `docs/` and publishes it

**Comment by fabd** [comment]
Out of curiosity I ran your fork locally.

It seems to work though I haven't checked any of the items stats etc.

You might want to update the "Feedback" url maybe to point to a reddit thread etc. You can change the full URL in [AppHeader line 40](https://github.com/kvothed2/pd2-runewizard/blob/main/src/components/AppHeader.vue#L40)

**Comment by kvothed2** [comment]
Hey thanks for very detailed response!
- Due to my ignorance on the topic, I haven't setup [Local Development](https://github.com/fabd/diablo2-runewizard#local-development), i may try to google and figure out what it means; i've just downloaded the repo and updated files in the editor
- I seem to grasp now that i actually need to build/compile the site (i thought that the repository is the site itself, and by making changes to content of it - it will be reflected)
- However, vite build, linux distro and nodejs packages, bash script - all of this is rather Chinese to me 😆 
- Currently, the site url is working, however it shows the old "build" of yours from that gh-pages/docs/assets/index.js file, since i haven't run the script to "remake" it

Not sure where to go from here. At the moment my fork is updated content-wise, but not built/compiled into the site.

**Comment by fabd** [comment]
Ok let's see

Hopefully you're on some kind of LINUX, or you can get some kind of linux Virtual Machine. Eg. VirtualBox + many ISOs are available.. this option would be the easiest on Windows even if it seems a bit complicated initially. VirutalBox is free and you download aan ISO and just install it and you can run the VM pretty easily. If you use a VM the only annoying part is you're going to need to set up a "shared folder" so you can edit your code outside of the VM.. but all of that is not Javascript programming so maybe you can figure it out.

So, assuming you're running some kind of linux ... open the "terminal" program.

**You can instal nodejs** like this , it's two lines :

https://github.com/nodesource/distributions#using-ubuntu

Don't worry too much about nodejs version, take a recent one.

This should install also **node package manager** which is the `npm `command.

So if you can get this part working, then in your terminal you should be able to do the Installation from this repo:

```
npm install
```

It will print a bunch of nonsense, all that matters is you now have a folder called 

    node_modules/

And this insane folder will be like 17000+ files for a total of 200 MB  ... welcome to Javascript !

---

So let me know if you can get this part working.

If YES, then just make sure you are in the root folder there, where `node_modules` is 

Then type  this to start your LOCAL server:

```
npm run dev
```

You will see this if all goes well:

```
vite v2.9.5 dev server running at:

  > Local: http://localhost:8080/diablo2-runewizard/
  > Network: use `--host` to expose

  ready in 291ms.
```

As you can see you can just go to that URL on your computer and you will see the site!

Let me know if yiu can get there - and can you use git locally ? Or at least have  `git` command work?

**Comment by fabd** [comment]
@kvothed2  You got me curious so I'm going to try with installing Ubuntu with "WSL" on WIndows 10. This might be easier than using VirtualBox.

Can you confirm what OS you use?

**Comment by kvothed2** [comment]
Hey @fabd yes i'm using Windows 10. 
I'm eager to try your suggested approach from above, although it will take some time. Will probably have to do it over next week (i didn't expect it is going to take that much time/efforts for a few hours of edits in the content). I downloaded Virtual Box, installed Python, and now need to update PATH for pip install win32 something to work lol (so that my virtual box installation works).
Once again, i'm not a web developer, nor programmer, so i go through all of this for the first time, and do not always understand things, so it will take some time 😄 Just wish there was an easier way lol

**Comment by fabd** [comment]
@kvothed2 Hold on. You got me curious and I'm trying to do it the Microsoft way. Using "Windows Subsystem for Linux" aka WSL 2 with Ubuntu. I think it will be a lot easier. I will update later.

In the meantime can you confirm what basic skills you have. Do you know basic use of git ? Did you do all the edits online with the github site? How did you switch to gh-pages etc?

**Comment by kvothed2** [comment]
I can make myself around github - commit, branch, push, merge etc
I downloaded zip of the repository, edited files with MS Visual Code and then uploaded files back with commits.
gh-pages were a second branch after the fork, i synched it with yours and made changes to it the same way as i did to main branch (uploaded edited files, made smaller changes online).
I worked before with multiple CMSes but more on the content/project manager side

**Comment by kvothed2** [comment]
So i've managed to get WSL2 and Ubuntu running, installed updates and upgraded, then got nodejs installed. Have the npm package, however
`npm install`
throws an error:
```
npm ERR! code ENOENT
npm ERR! syscall open
npm ERR! path /home/kvothed/package.json
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, open '/home/kvothed/package.json'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/kvothed/.npm/_logs/2022-11-24T22_56_51_751Z-debug-0.log
```
Perhaps, i'm doing this command in a wrong place, but not sure how to get wherever i need. I'm not very familiar with terminal/command line to navigate there.

**Comment by kvothed2** [comment]
Ok, more progress :D
I cloned my repository with 
```
git clone https://github.com/kvothed2/pd2-runewizard.git
```
and ran
```
npm install
npm run dev
```
got my site fully working on 
http://localhost:8080/pd2-runewizard/

Everything looks as expected (minor css issue tho). I will continue checking things. Thanks for all the comments you left -they help a lot :D

**Comment by fabd** [comment]
Ahh congrats, I was busy writing a step by step.

It's actually interesting to me , as working with WSL 2 was an option before I switched to Ubuntu natively.

So, for nodejs initially it didn't work. `sudo apt install nodejs npm` installs an old version of node (v10)

It's better to install the nodesource version here https://github.com/nodesource/distributions#using-ubuntu

I'm guessing that's what you did? In this case I get:

```bash
$ type node
node is /usr/bin/node
$ node -v
v19.1.0
$ npm -v
8.19.3
 ```

For Visual Studio Code, you install the "WSL" addon.

Then from the terminal you type just `code .` to open the folder, pretty cool.

You can see in VS Code side bar then, that your files are here :  `\\wsl$\Ubuntu\home\fabd\pd2-runewizard`  (change the username)

So you can also pin your Ubuntu home folder if you like in File Explorer and it's safe to access files that way, to gain a better understanding of all the folders.

Now I'm trying to figure out a GUI friendly way to update the git repo.

I installed Sublime Merge - which I like - but it doesn't seem to work will with WSL.

**Comment by fabd** [comment]
You can also install Windows Terminal for a better experience.

Here's what I had in my notes so far:

---

**Install Windows Terminal from the Windows Store**

It's wayyyyy nicer than PowerShell, for typing commands.

Click "Open" to start it.

Go to Settings, right now you might want to change:

 - Startup > Default profile > set "Ubuntu"
 - Profiles > Ubuntu > Color scheme & Font size

   ... you might want to "Pin to taskbar" as well at this point.

   Note how you can RIGHT CLICK to paste any text, very useful.

   Basic terminal shortcuts:
      ctrl left/right  to move cursor
      ctrl a    ... start of line
      ctrl k    ... delete the line

    Note how you can use CURSOR UP/DOWN to repeat previous commands!

**Comment by kvothed2** [comment]
Hey, I used the nodesource command that you posted before, now i get same node, npm version as you wrote, and type node is hashed.
I did get Windows Terminal, it is much better that way, definitely. Also WSL addon to VS Code is nice that is true :) 

Now at this point few things are still not clear:
- Where do i need to do changes in order to update the local site? In \\wsl$\Ubuntu\home\kvothed\pd2-runewizard? or back in GitHub Main branch and then "fetch" it into my local above mentioned respository?
- I assume i can push updates if i do them locally back into github?
- if above is correct, then i can finalize and iron out a few things; how do i perform then your bash script to update gh-pages and compile the site?

Do you think i'm missing any steps in between?

**Comment by fabd** [comment]
Yes, you make changes to your local repo - it is after all the "source of truth".

Use VSCode or any editor you like.  With VSCode you can just `cd ` in the repo folder, and then type `code .` and then work from there.  If you eg. used Notepad then you'd use that WSL path `\\wsl$\Ubuntu\home\kvothed\pd2-runewizard`.

> I assume i can push updates if i do them locally back into github?

Yes.

>  how do i perform then your bash script to update gh-pages and compile the site?

We're not done yet I'm afraid.

We need to push the changes whether it is on main branch or gh-pages branch.

I'm still trying to figure that out.

I like **Sublime Merge** as a GUI, but it doesn't seem to work well with the repo inside WSL 2.

### GIT setup

You'll need at very least to set your git user info  - make sure it matches your github profile

```
$ git config --global user.email  your.name@gmail.com
$ git config --global user.name kvothed2
```

**Comment by fabd** [comment]
Without a GUI pushing git changes is a real PITA.

### First you need to set your author info locally , make sure it matches Github profile

```
$ git config --global user.email  your.name@gmail.com
$ git config --global user.name kvothed2
```

You can check that gitconfig it's in the root folder `.gitconfig` file editable in notepad as well

### Next push changes

With command line it's a PITA , but this should work at this stage:

(you edit files)

You confirm modified files with 

    git status
 
You can see changes with

    git diff

You add changes with 

    git add .

You commit with

    git commit -m "test commit title"

You push 

    git push

Whew

Try this first. I may have to go to bed.

Note: _ALWAYS commit first before switching branch or you may lose changes! _

To create a branch and "checkout" to it (switch to it) in one command :

    git checkout -b gh-pages


To just checkout / switch back to eg. main : 

    git checkout main

So again

    git checkout gh-pages

Now we need to tell git how this local branch mathces the one online , otherwise we can't push :  
    
    git branch --set-upstream-to=origin/gh-pages gh-pages

As you see it says something like "gh-pages should be pushed to origin/gh-pages (origin being the corresponding repo, the one you cloned from github)

Build the site

    npm run build

Now to make things simple let's just delete any previous docs/ folder, and replace it by simply renaming the build folder : 

    rm -rf docs/
    mv dist/ docs


So we rename the vite output "dist" into "docs" got Github Pages

Add the whole folder to the "stage",  commit, and push

    git commit docs/
    git commit -m "test deployment"
    git push -f

Note the `-f` to force replacing whatever gh-pages is at on the remote - otherwise it will complain.

Don't forget you are on `gh-pages` and go back to master  for editing

    git checkout main

To check your current branches

    git branch -v

To repeat the publish process, let's just DELETE the gh-pages and re-make it so we don't care about merging etc.

    git checkout -b gh-pages
    npm run build
    git rm -rf docs ; git mv dist docs
    git add docs
    git commit -m "this is so much fun"
    git push -f

You could put that in a small "deploy" script if you like. Here is an example of creating a small script and running it:

    echo "echo hi am running from a script" > test.sh
    chmod +x test.sh
    ./test.sh

Note how we type `./test.sh` not `test.sh`

All you'd want at this point is a git GUI so it's a lot easier to review your changes, commit them, and push/pull/merge stuff;

I'm off to bed - if you find a GUI that works let me know.

**Comment by fabd** [comment]
Update fixing the repeatable deploy steps that you can put in a script : 

To repeat the publish process, let's just DELETE the gh-pages and re-make it so we don't care about merging etc.

    git branch -D gh-pages
    git checkout -b gh-pages
    git branch --set-upstream-to=origin/gh-pages gh-pages
    npm run build
    rm -rf docs ; mv dist docs
    git add docs
    git commit -m "this is so much fun"
    git push -f

The first line will fail if you forgot you're already in gh-pages, which is fine.

The second line won't do anything if we're already in gh-pages which is also fine.

Third line is because if we deleted the branch we have to "set the upstream again"



You could put that in a small "deploy" script if you like. Here is an example of creating a small script and running it:

    echo "echo hi am running from a script" > test.sh
    chmod +x test.sh
    ./test.sh

Note how we type `./test.sh` not `test.sh`

All you'd want at this point is a git GUI so it's a lot easier to review your changes, commit them, and push/pull/merge stuff;

Let me know if you got the gh-pages publish working. Signing out for real  :P

**Comment by fabd** [comment]
@kvothed2 

You can do GIT stuff inside VSCode with some of the extensions - but I don't know it handles git itself, whether you need the Windows version of git.

It sounds like running native linux apps with GUI could work, though I haven't got time to try right now - let me know if you did.

https://learn.microsoft.com/en-us/windows/wsl/tutorials/gui-apps

In that case perhaps the following might just work?

gitg is a simple small GUI for doing common git operations.

```
sudo apt install gitg
gitg
```

**Comment by kvothed2** [comment]
Hey @fabd I actually just ran my commits and pushed them to main, right from the VSCode interface, just needed to resolve one Merge conflict and it all worked! 
So i will finalize the changes, push them again to main and then look at your latest instructions on how to compile the build into gh-pages! Hopefully in an hour I can share working URL 😆 
Thank you!

**Comment by fabd** [comment]
Sounds good!

Actually I have no idea how to do the git stuff inside of VSCode, did you use an exension? And/or did you install git in Windows separately? I'm gonna have to try that later.

**Comment by fabd** [comment]
PS: if you make some small imrpvoements here and there or even interesting visual tweaks, try to make separate smaller commits with just css/html.. maybe I'll be interested in merging some back. Design wise,  it's quite hard to make a good looking table!

**Comment by kvothed2** [comment]
There it is 😆 

https://kvothed2.github.io/pd2-runewizard/

In the end i had denied permissions to move dist into docs, so i `cp -r dist/ docs` and it worked. 
OMG, Thank you so much! Without your directions and help it wouldn't be possible!

There is a Source Control in VSCode left side bar, that allowed to commit/push/pull/synch. All the merge conflicts were also showcased and i could resolve them there with the editor. 
Now one thing that i probably messed up and need to figure out "how to" is that when I wanted to push my gh-pages changes, i first had to pull from origin, that can cause some merge conflicts and /or difference with main. Is there a way to "replace" origin with local? 
I imagine next time i work on main, delete and re-create gh-pages branch and will encounter the same issue that it is not the same between origin and local.
Not sure if that is clear what the issue is?
Thank you anyways, it is time to QA and get ready for the Season 6 start in just 2 hours. Can't believe we made it before launch lol

**Comment by kvothed2** [comment]
Now i will need to write small cheat sheet for myself in confluence to remember how to maintain all of that :D

**Comment by fabd** [comment]
Hah, nice!

You're welcome - it was fun seeing what I can do in Windows starting from scratch. Also didn't try WSL in a long time.

I like that you styled it a bit, gives it an identity and then you can immediately tell its a different version.

Yeah, we don't really need to move `dist/` or even remove it - it's ignored by `.gitignore` anyway so it can just hang around.

There's many ways to handle gh-pages. Could have also committed docs/ to the `main` branch, and merge main into gh-pages repeatedly, instead of resetting it. Either way works. Just that docs/ is compiled code, and it may or may not be up to date with the actual source CSS/JS so I feel like it adds confusion - so the `main` branch I set up doesn't have any "built" files in it.

You shouldn't have to pull - you could do just the last "push" step on the command line. Do the commits etc in VSCode then do `git push -f` on the command line. Make sure to check which branch is active first `git branch -v` . VSCode probably has a "force push" flag somewhere when you do the push (I don't know what it looks like, haven't tried).

> Not sure if that is clear what the issue is?

Yes you got it, `gh-pages` is a weird branch - and in fact in some BIG repos out there you'll see they even have an entirely different file tree. Like all the code disappears, and a bunch of html docs appear instead. But personally I don't like having my filesystem delete and create a ton of files like that when I switch branch ... just feels wrong .. and so I preferred the approach of just adding the `docs/` folder in gh-pages , because if you checkout back into `main` git doesn't complain because all the other files haven't changed, AND the `docs/` folder  - _this is crucial point by the way_ - the docs/ folder does NOT exist in `main`- so you don't get any git conflicts if you temporarily checkout into main while the docs/ folder is there. Likewise for dist/ it doesn't get committed in gh-pages either (it's ignored in both branches). So TLDR it's less headaches when you have these two folders dist/ and docs/ that have no conflicts when you change branches.

More to the point .. I suppose if you wanted to, you could commit (so ADD) the docs/ folder in the gh-pages. THEN when you do a new build `npm run build` you'd COPY over the docs/ files. This would cause git to see differences, so you'd commit those differences (basically all the compiled CSS/JS files changed). And then you'd commit that, push. And then what that means is your gh-pages has a long running history which doesn't get broken, it continues forward in time if that makes sense? So in that case your local version will always be ahead by 1+ commits, and you get no conflicts. 

In other words when we reset the branch we reset the first commit HASH in that branch, therefore it appears as completely different, and git complains it doesn't match the remote one. But since we know that we just "force push".

Glad you got it working!

I'll close the thread as it is resolved but of course you can still post in it if you have another question - cheers!

**Comment by fabd** [comment]
> In other words when we reset the branch we reset the first commit HASH in that branch, therefore it appears as completely different, and git complains it doesn't match the remote one. But since we know that we just "force push".

Actually I'm slightly incorrect there. When the gh-pages branch is deleted and recreated - the LAST commit there you make after you do a new build, it doesn't match the one on the remote. So the history is broken. Like A > B > C > LOL locally, and A > B > C > CAT  on remote. git expects that your local branch has the same history so it wants you to first pull LOL , then commit on top of it. eg. A > B > C > LOL > CAT - which you can then push to remote without conflict since now you are ahead by one commit.

Aaanyways - have fun , haven't been able to play D2R yet!
